drivers/dma/dma_dw_axi.c

```diff
@@ -0,0 +1,912 @@
+/*
+ * Copyright (c) 2023 Intel Corporation.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#define DT_DRV_COMPAT snps_designware_dma_axi
+
+#include <zephyr/device.h>
+#include <zephyr/drivers/dma.h>
+#include <zephyr/logging/log.h>
+#include <zephyr/drivers/reset.h>
+#include <zephyr/cache.h>
+
+LOG_MODULE_REGISTER(dma_designware_axi, CONFIG_DMA_LOG_LEVEL);
+
+#define DEV_CFG(_dev)	((const struct dma_dw_axi_dev_cfg *)(_dev)->config)
```

Comment:
**Conflict Requirement**

There is one conflict requirement of struct dma_context(from dma.h) and DEVICE_MMIO_RAM(from device_mmio.h) to be a first member of a driver data structure. 

1. 
dma.h expects dma_context structure to be a first member of driver struct data:
```
/**
 * DMA context structure
 * Note: the dma_context shall be the first member
 *       of DMA client driver Data, got by dev->data
 */
struct dma_context {
	/** magic code to identify the context */
	int32_t magic;
	/** number of dma channels */
	int dma_channels;
	/** atomic holding bit flags for each channel to mark as used/unused */
	atomic_t *atomic;
};
```

2. 
DEVICE_MMIO_RAM: 
snippet from device_mmio.h
` * struct foo_driver_data {
 *	DEVICE_MMIO_RAM;       -->expected as first member of driver data structure
 *	int wibble;
 *	...
 * }`

**Solution:**
So Since this is a conflict requirement and these both conditions can't be satisfied at the same time. We have other variation of DEIVICE_MMIO_* called DEVICE_MMIO_NAMED_*.

These DEVICE_MMIO_NAMED* macros can use the struct type of the driver's data structure to access the virtual address stored in struct data. So decided to use DEVICE_MMIO_NAMED_* macros instead of DEVICE_MMIO_* macros

As we have to use the DEVICE_MMIO_NAMED* macros, device_mmio.h internally expects DEV_CFG and DEV_DATA to be defined by driver so that device_mmio.h can access the proper address by typecasting